chapters/part3/hyperparameter-optimization-with-optuna.md

```diff
@@ -0,0 +1,419 @@
+
+```python
+optuna.visualization.plot_optimization_history(study)
+```
+
+<br>
+<img src="/part3/hyperparameter-optimization-with-optuna/optimization_history.jpg" alt="Plot Optimization History" />
+
+From the recent release of Optuna, we can evaluate parameter importances based on finished trials.
+There are two evaluators available: the default [fANOVA](http://proceedings.mlr.press/v32/hutter14.html) and [MDI](https://papers.nips.cc/paper/4928-understanding-variable-importances-in-forests-of-randomized-trees).
+To show importances of hyperparameters, it uses `optuna.visualization.plot_param_importances`:
+
+```python
+optuna.visualization.plot_param_importances(study)
+```
+
+In this plot, we can see that `lr` is the most important hyperparameter in this experiment.
+
+<br>
+<img src="/part3/hyperparameter-optimization-with-optuna/hyperparameter_importance.jpg" alt="Plot Hyperparameter Importance" />
+
+Additionally, you can export a configuration with optimized hyperparameters.
+
+```python
+dump_best_config("./imdb_optuna.jsonnet", "./best_config.json", study)
+```
+
+It will create a configuration named `best_config.json`.
+This is helpful to retrain a model with the best hyperparameters.
+
+</exercise>
+
+<exercise id="4" title="Writing your own script with Optuna and Use Pruners">
+
+Additionally, you can also use Optuna when you define the AllenNLP model by writing your script.
+A python code of training a model with AllenNLP typically consists of main three parts: preparing data, building model, and creating trainer.
+<!-- We put the python code to use AllenNLP+Optuna at the bottom of this exercise. -->
+
+## Pruning unpromising trials
+
+Hyperparameter optimization often takes a long time to find good parameters.
+If you can find and stop unpromising trials with bad hyperparameters, you can reduce the time of optimization and get good hyperparameters faster.
+Stopping unpromising trials is so-called pruning.
+The following illustration shows an example of pruning.
+Although the final curves at the left side of the picture are not available before finishing hyperparameter optimization,
+a pruner evaluates at each epoch how promising each trial will be and stops if it predicts to be unpromising.
+
+<img src="/part3/hyperparameter-optimization-with-optuna/illustration_of_pruning.jpg" alt="Illustration of Pruning" />
+
+You can use `AllenNLPPruningCallback` that is the new feature of Optuna, which allows users to prune unpromising trials with algorithms implemented in Optuna.
+`AllenNLPPruningCallback` is the interface to provide the way to use these pruning algorithms in `GradientDescentTrainer`, which is the standard way to train a model in AllenNLP.
+To enable a pruning, you must create the AllenNLP callback first. `AllenNLPPruningCallback` needs two arguments: `trial` and a target metric.
```

Comment:
"`trial` and a target metric' is a bit inconsistent. How about removing the backticks of the trial?
```suggestion
To enable a pruning, you must create the AllenNLP callback first. `AllenNLPPruningCallback` needs two arguments: a trial and a target metric.
```